ecs-agent/api/resource/resource_attachment.go

```diff
@@ -83,3 +96,97 @@ func getExtensibleEphemeralStorageProperties() (ephemeralStorageProperties []str
 	ephemeralStorageProperties = append(ephemeralStorageProperties, getFargateControlPlaneProperties()...)
 	return ephemeralStorageProperties
 }
+
+func getResourceAttachmentLogFields(ra *ResourceAttachment, duration time.Duration) logger.Fields {
+	fields := logger.Fields{
+		"duration":          duration.String(),
+		"attachmentARN":     ra.AttachmentARN,
+		"attachmentType":    ra.AttachmentProperties[ResourceTypeName],
+		"attachmentSent":    ra.AttachStatusSent,
+		"volumeSizeInGiB":   ra.AttachmentProperties[VolumeSizeInGiBName],
+		"RequestedSizeName": ra.AttachmentProperties[RequestedSizeName],
+		"volumeId":          ra.AttachmentProperties[VolumeIdName],
+		"deviceName":        ra.AttachmentProperties[DeviceName],
+		"filesystemType":    ra.AttachmentProperties[FileSystemTypeName],
+		"status":            ra.Status.String(),
+		"expiresAt":         ra.ExpiresAt.Format(time.RFC3339),
+	}
+
+	return fields
+}
+
+// StartTimer starts the ack timer to record the expiration of resource attachment
+func (ra *ResourceAttachment) StartTimer(timeoutFunc func()) error {
+	ra.guard.Lock()
+	defer ra.guard.Unlock()
+
+	if ra.ackTimer != nil {
+		// The timer has already been initialized, do nothing
+		return nil
+	}
+	now := time.Now()
+	duration := ra.ExpiresAt.Sub(now)
+	if duration <= 0 {
+		return errors.Errorf("resource attachment: timer expiration is in the past; expiration [%s] < now [%s]",
+			ra.ExpiresAt.String(), now.String())
+	}
+	logger.Info("Starting resource attachment ack timer", getResourceAttachmentLogFields(ra, duration))
+	ra.ackTimer = time.AfterFunc(duration, timeoutFunc)
+	return nil
+}
+
+// Initialize initializes the fields that can't be populated from loading state file.
+// Notably, this initializes the ack timer so that if we time out waiting for the resource to be attached, the attachment
+// can be removed from state.
+func (ra *ResourceAttachment) Initialize(timeoutFunc func()) error {
+	ra.guard.Lock()
+	defer ra.guard.Unlock()
+
+	if ra.AttachStatusSent { // resource attachment status has been sent, no need to start ack timer.
+		return nil
+	}
+
+	now := time.Now()
+	duration := ra.ExpiresAt.Sub(now)
+	if duration <= 0 {
+		return errors.New("Resource attachment has already expired")
```

Comment:
nit: go convention is to not capitalize first character of error strings

Reply:
fixed